--- a/src/devices/wifi/nm-device-wifi.c
+++ b/src/devices/wifi/nm-device-wifi.c
@@ -1764,17 +1764,10 @@
 		nm_act_request_clear_secrets (req);
 
 		_LOGI (LOGD_DEVICE | LOGD_WIFI,
-		       "Activation: (wifi) disconnected during association, asking for new key");
+		       "Activation: (wifi) disconnected during association");
 
 		cleanup_association_attempt (self, TRUE);
 		nm_device_state_changed (device, NM_DEVICE_STATE_NEED_AUTH, NM_DEVICE_STATE_REASON_SUPPLICANT_DISCONNECT);
-		nm_act_request_get_secrets (req,
-		                            setting_name,
-		                            NM_SECRET_AGENT_GET_SECRETS_FLAG_ALLOW_INTERACTION
-		                              | NM_SECRET_AGENT_GET_SECRETS_FLAG_REQUEST_NEW,
-		                            NULL,
-		                            wifi_secrets_cb,
-		                            self);
 		handled = TRUE;
 	}
 
@@ -2056,24 +2049,24 @@
 	applied_connection = nm_act_request_get_applied_connection (req);
 
 	tries = GPOINTER_TO_UINT (g_object_get_data (G_OBJECT (applied_connection), WIRELESS_SECRETS_TRIES));
-	if (tries > 3)
+	if (tries > 5 || new_secrets) {
+		nm_act_request_clear_secrets (req);
+		setting_name = nm_connection_need_secrets (applied_connection, NULL);
+		if (setting_name) {
+			NMSecretAgentGetSecretsFlags flags = NM_SECRET_AGENT_GET_SECRETS_FLAG_ALLOW_INTERACTION;
+
+			if (new_secrets)
+				flags |= NM_SECRET_AGENT_GET_SECRETS_FLAG_REQUEST_NEW;
+			nm_act_request_get_secrets (req, setting_name, flags, NULL, wifi_secrets_cb, self);
+		} else
+			_LOGW (LOGD_DEVICE, "Cleared secrets, but setting didn't need any secrets.");
+
 		return NM_ACT_STAGE_RETURN_FAILURE;
+	}
 
 	nm_device_state_changed (NM_DEVICE (self), NM_DEVICE_STATE_NEED_AUTH, NM_DEVICE_STATE_REASON_NONE);
-
-	nm_act_request_clear_secrets (req);
-	setting_name = nm_connection_need_secrets (applied_connection, NULL);
-	if (setting_name) {
-		NMSecretAgentGetSecretsFlags flags = NM_SECRET_AGENT_GET_SECRETS_FLAG_ALLOW_INTERACTION;
-
-		if (new_secrets)
-			flags |= NM_SECRET_AGENT_GET_SECRETS_FLAG_REQUEST_NEW;
-		nm_act_request_get_secrets (req, setting_name, flags, NULL, wifi_secrets_cb, self);
-
-		g_object_set_data (G_OBJECT (applied_connection), WIRELESS_SECRETS_TRIES, GUINT_TO_POINTER (++tries));
-		ret = NM_ACT_STAGE_RETURN_POSTPONE;
-	} else
-		_LOGW (LOGD_DEVICE, "Cleared secrets, but setting didn't need any secrets.");
+	g_object_set_data (G_OBJECT (applied_connection), WIRELESS_SECRETS_TRIES, GUINT_TO_POINTER (++tries));
+	ret = NM_ACT_STAGE_RETURN_POSTPONE;
 
 	return ret;
 }
